File: packages/sdk-java/harness-sdk-core/src/main/java/com/harness/core/ErrorHandler.java
package com.harness.core;

import java.util.HashMap;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Intelligent error handler with recovery strategies.
 *
 * Maps error types to appropriate recovery actions:
 * - RateLimitError → RETRY with exponential backoff
 * - ContextTooLongError → COMPRESS_CONTEXT
 * - PermissionDeniedError → ABORT
 * - TimeoutError → RETRY (max 3 times)
 * - NetworkError → RETRY with backoff
 * - ToolExecutionError → SKIP or RETRY
 *
 * Example:
 * <pre>
 * ErrorHandler handler = new ErrorHandler();
 * ErrorDecision decision = handler.handle(error, context);
 * if (decision.action() == ErrorAction.RETRY) {
 *     Thread.sleep((long) (decision.delaySeconds() * 1000));
 *     // retry the operation
 * }
 * </pre>
 */
public class ErrorHandler {

    private static final Logger logger = LoggerFactory.getLogger(ErrorHandler.class);

    private final int maxRetries;
    private final double baseDelay;
    private final double maxDelay;

    // Track retry attempts per operation
    private final Map<String, Integer> retryCounts = new HashMap<>();

    public ErrorHandler() {
        this(3, 1.0, 60.0);
    }

    public ErrorHandler(int maxRetries, double baseDelay, double maxDelay) {
        this.maxRetries = maxRetries;
        this.baseDelay = baseDelay;
        this.maxDelay = maxDelay;
    }

    /**
     * Determine how to handle an error.
     *
     * @param error The exception that occurred
     * @param context Context about the error situation
     * @return ErrorDecision with action and parameters
     */
    public ErrorDecision handle(Exception error, ErrorContext context) {
        String errorType = error.getClass().getSimpleName();
        String errorMessage = error.getMessage() != null ? error.getMessage().toLowerCase() : "";

        // Check for rate limit errors
        if (isRateLimitError(errorType, errorMessage)) {
            return handleRateLimit(context);
        }

        // Check for context overflow
        if (isContextError(errorType, errorMessage, context)) {
            return handleContextOverflow(context);
        }

        // Check for permission errors
        if (isPermissionError(errorType, errorMessage)) {
            return handlePermissionError(error, context);
        }

        // Check for timeout errors
        if (isTimeoutError(errorType, errorMessage)) {
            return handleTimeout(context);
        }

        // Check for network errors
        if (isNetworkError(errorType, errorMessage)) {
            return handleNetworkError(context);
        }

        // Check for tool errors
        if (context.toolName() != null) {
            return handleToolError(error, context);
        }

        // Default: abort on unknown errors
        return ErrorDecision.abort("Unhandled error type: " + errorType + ": " + error.getMessage());
    }

    /**
     * Reset retry counters.
     */
    public void reset() {
        retryCounts.clear();
    }

    // === Error Type Detection ===

    private boolean isRateLimitError(String errorType, String message) {
        String[] indicators = {"ratelimit", "rate_limit", "429", "too many requests", "rate limit", "quota exceeded"};
        return errorType.toLowerCase().contains("rate") ||
               containsAny(message, indicators);
    }

    private boolean isContextError(String errorType, String message, ErrorContext context) {
        if (context.isContextOverflow()) {
            return true;
        }
        String[] indicators = {"context", "token limit", "max_tokens", "too long", "length", "context_length_exceeded"};
        return containsAny(message, indicators);
    }

    private boolean isPermissionError(String errorType, String message) {
        String[] indicators = {"permission", "forbidden", "403", "unauthorized", "access denied", "not allowed"};
        return errorType.toLowerCase().contains("permission") ||
               containsAny(message, indicators);
    }

    private boolean isTimeoutError(String errorType, String message) {
        String[] indicators = {"timeout", "timed out", "deadline"};
        return errorType.toLowerCase().contains("timeout") ||
               containsAny(message, indicators);
    }

    private boolean isNetworkError(String errorType, String message) {
        String[] typeIndicators = {"connection", "network", "socket"};
        String[] messageIndicators = {"connection", "network", "socket", "dns", "refused", "unreachable"};

        return containsAny(errorType.toLowerCase(), typeIndicators) ||
               containsAny(message, messageIndicators);
    }

    private boolean containsAny(String text, String[] patterns) {
        for (String pattern : patterns) {
            if (text.contains(pattern)) {
                return true;
            }
        }
        return false;
    }

    // === Error Handlers ===

    private ErrorDecision handleRateLimit(ErrorContext context) {
        String key = "ratelimit_" + context.iteration();
        int attempts = retryCounts.getOrDefault(key, 0);

        if (attempts >= maxRetries) {
            return ErrorDecision.abort("Rate limit persisted after " + attempts + " retries");
        }

        // Exponential backoff
        double delay = Math.min(baseDelay * Math.pow(2, attempts), maxDelay);
        retryCounts.put(key, attempts + 1);

        logger.warn("Rate limited, waiting {}s before retry (attempt {})", delay, attempts + 1);
        return ErrorDecision.retry(delay, "Rate limited, waiting " + String.format("%.1f", delay) + "s before retry");
    }

    private ErrorDecision handleContextOverflow(ErrorContext context) {
        if (context.iteration() > 3) {
            return ErrorDecision.abort("Context overflow persists after compression attempts");
        }

        logger.warn("Context too long ({} tokens), attempting compression", context.contextTokens());
        return ErrorDecision.compressContext("Context too long, attempting compression");
    }

    private ErrorDecision handlePermissionError(Exception error, ErrorContext context) {
        logger.error("Permission denied: {}", error.getMessage());
        return ErrorDecision.abort("Permission denied: " + error.getMessage());
    }

    private ErrorDecision handleTimeout(ErrorContext context) {
        String key = "timeout_" + context.iteration();
        int attempts = retryCounts.getOrDefault(key, 0);

        if (attempts >= maxRetries) {
            return ErrorDecision.abort("Operation timed out after " + attempts + " retries");
        }

        double delay = Math.min(baseDelay * Math.pow(2, attempts), maxDelay);
        retryCounts.put(key, attempts + 1);

        logger.warn("Timeout, retrying in {}s (attempt {})", delay, attempts + 1);
        return ErrorDecision.retry(delay, "Timeout, retrying in " + String.format("%.1f", delay) + "s");
    }

    private ErrorDecision handleNetworkError(ErrorContext context) {
        String key = "network_" + context.iteration();
        int attempts = retryCounts.getOrDefault(key, 0);

        if (attempts >= maxRetries) {
            return ErrorDecision.abort("Network error persisted after " + attempts + " retries");
        }

        double delay = Math.min(baseDelay * Math.pow(2, attempts), maxDelay);
        retryCounts.put(key, attempts + 1);

        logger.warn("Network error, retrying in {}s (attempt {})", delay, attempts + 1);
        return ErrorDecision.retry(delay, "Network error, retrying in " + String.format("%.1f", delay) + "s");
    }

    private ErrorDecision handleToolError(Exception error, ErrorContext context) {
        String errorMessage = error.getMessage() != null ? error.getMessage().toLowerCase() : "";

        // Some tool errors are recoverable
        if (errorMessage.contains("not found") || errorMessage.contains("does not exist")) {
            return ErrorDecision.skip("Tool " + context.toolName() + ": resource not found");
        }

        if (errorMessage.contains("invalid") || errorMessage.contains("invalid argument")) {
            // Bad arguments, don't retry
            return ErrorDecision.skip("Tool " + context.toolName() + ": invalid arguments");
        }

        // Other tool errors: try once more
        String key = "tool_" + context.toolName() + "_" + context.iteration();
        int attempts = retryCounts.getOrDefault(key, 0);

        if (attempts >= 1) {
            return ErrorDecision.skip("Tool " + context.toolName() + " failed, skipping");
        }

        retryCounts.put(key, attempts + 1);
        logger.warn("Tool {} failed, retrying", context.toolName());
        return ErrorDecision.retry(baseDelay, "Tool " + context.toolName() + " failed, retrying");
    }
}

File: packages/sdk/src/harness/core/error_handler.py
"""
Error handling strategy for agent loop.

Provides intelligent error recovery based on error type.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any


class ErrorAction(Enum):
    """Actions to take when an error occurs."""

    RETRY = "retry"  # Retry the operation
    COMPRESS_CONTEXT = "compress"  # Compress context and retry
    ABORT = "abort"  # Stop execution
    SKIP = "skip"  # Skip and continue
    ESCALATE = "escalate"  # Escalate to user


@dataclass
class ErrorContext:
    """Context for error handling decision."""

    error: Exception
    iteration: int
    tool_name: str | None = None
    attempt: int = 1
    context_tokens: int = 0
    max_tokens: int = 200000
    previous_errors: list[Exception] | None = None

    @property
    def is_context_overflow(self) -> bool:
        """Check if error is due to context overflow."""
        return self.context_tokens > self.max_tokens * 0.9


@dataclass
class ErrorDecision:
    """Decision on how to handle an error."""

    action: ErrorAction
    delay_seconds: float = 0.0
    message: str = ""
    metadata: dict[str, Any] | None = None


class ErrorHandler:
    """
    Intelligent error handler with recovery strategies.

    Maps error types to appropriate recovery actions:
    - RateLimitError → RETRY with exponential backoff
    - ContextTooLongError → COMPRESS_CONTEXT
    - PermissionDeniedError → ABORT
    - TimeoutError → RETRY (max 3 times)
    - NetworkError → RETRY with backoff
    - ToolExecutionError → SKIP or RETRY

    Example:
        >>> handler = ErrorHandler()
        >>> decision = handler.handle(error, context)
        >>> if decision.action == ErrorAction.RETRY:
        ...     await asyncio.sleep(decision.delay_seconds)
        ...     # retry the operation
    """

    def __init__(
        self,
        max_retries: int = 3,
        base_delay: float = 1.0,
        max_delay: float = 60.0,
    ):
        self.max_retries = max_retries
        self.base_delay = base_delay
        self.max_delay = max_delay

        # Track retry attempts per operation
        self._retry_counts: dict[str, int] = {}

    def handle(self, error: Exception, context: ErrorContext) -> ErrorDecision:
        """
        Determine how to handle an error.

        Args:
            error: The exception that occurred
            context: Context about the error situation

        Returns:
            ErrorDecision with action and parameters
        """
        error_type = type(error).__name__
        error_message = str(error).lower()

        # Check for rate limit errors
        if self._is_rate_limit_error(error_type, error_message):
            return self._handle_rate_limit(context)

        # Check for context overflow
        if self._is_context_error(error_type, error_message, context):
            return self._handle_context_overflow(context)

        # Check for permission errors
        if self._is_permission_error(error_type, error_message):
            return self._handle_permission_error(error, context)

        # Check for timeout errors
        if self._is_timeout_error(error_type, error_message):
            return self._handle_timeout(context)

        # Check for network errors
        if self._is_network_error(error_type, error_message):
            return self._handle_network_error(context)

        # Check for tool errors
        if context.tool_name:
            return self._handle_tool_error(error, context)

        # Default: abort on unknown errors
        return ErrorDecision(
            action=ErrorAction.ABORT,
            message=f"Unhandled error type: {error_type}: {error}",
        )

    def _is_rate_limit_error(self, error_type: str, message: str) -> bool:
        """Check if this is a rate limit error."""
        rate_limit_indicators = [
            "ratelimit",
            "rate_limit",
            "429",
            "too many requests",
            "rate limit",
            "quota exceeded",
        ]
        return "rate" in error_type.lower() or any(
            indicator in message for indicator in rate_limit_indicators
        )

    def _is_context_error(
        self,
        error_type: str,
        message: str,
        context: ErrorContext,
    ) -> bool:
        """Check if this is a context length error."""
        context_indicators = [
            "context",
            "token limit",
            "max_tokens",
            "too long",
            "length",
            "context_length_exceeded",
        ]
        return context.is_context_overflow or any(
            indicator in message for indicator in context_indicators
        )

    def _is_permission_error(self, error_type: str, message: str) -> bool:
        """Check if this is a permission denied error."""
        permission_indicators = [
            "permission",
            "forbidden",
            "403",
            "unauthorized",
            "access denied",
            "not allowed",
        ]
        return "permission" in error_type.lower() or any(
            indicator in message for indicator in permission_indicators
        )

    def _is_timeout_error(self, error_type: str, message: str) -> bool:
        """Check if this is a timeout error."""
        timeout_indicators = [
            "timeout",
            "timed out",
            "deadline",
        ]
        return "timeout" in error_type.lower() or any(
            indicator in message for indicator in timeout_indicators
        )

    def _is_network_error(self, error_type: str, message: str) -> bool:
        """Check if this is a network error."""
        network_indicators = [
            "connection",
            "network",
            "socket",
            "dns",
            "refused",
            "unreachable",
        ]
        return any(
            term in error_type.lower() for term in ["connection", "network", "socket"]
        ) or any(indicator in message for indicator in network_indicators)

    def _handle_rate_limit(self, context: ErrorContext) -> ErrorDecision:
        """Handle rate limit errors with exponential backoff."""
        key = f"ratelimit_{context.iteration}"
        attempts = self._retry_counts.get(key, 0)

        if attempts >= self.max_retries:
            return ErrorDecision(
                action=ErrorAction.ABORT,
                message=f"Rate limit persisted after {attempts} retries",
            )

        # Exponential backoff
        delay = min(self.base_delay * (2**attempts), self.max_delay)
        self._retry_counts[key] = attempts + 1

        return ErrorDecision(
            action=ErrorAction.RETRY,
            delay_seconds=delay,
            message=f"Rate limited, waiting {delay:.1f}s before retry (attempt {attempts + 1})",
        )

    def _handle_context_overflow(self, context: ErrorContext) -> ErrorDecision:
        """Handle context length errors."""
        if context.iteration > 3:
            # Already tried compression multiple times
            return ErrorDecision(
                action=ErrorAction.ABORT,
                message="Context overflow persists after compression attempts",
            )

        return ErrorDecision(
            action=ErrorAction.COMPRESS_CONTEXT,
            message="Context too long, attempting compression",
            metadata={"current_tokens": context.context_tokens},
        )

    def _handle_permission_error(
        self,
        error: Exception,
        context: ErrorContext,
    ) -> ErrorDecision:
        """Handle permission denied errors."""
        return ErrorDecision(
            action=ErrorAction.ABORT,
            message=f"Permission denied: {error}",
            metadata={"tool": context.tool_name},
        )

    def _handle_timeout(self, context: ErrorContext) -> ErrorDecision:
        """Handle timeout errors with retry."""
        key = f"timeout_{context.iteration}"
        attempts = self._retry_counts.get(key, 0)

        if attempts >= self.max_retries:
            return ErrorDecision(
                action=ErrorAction.ABORT,
                message=f"Operation timed out after {attempts} retries",
            )

        delay = min(self.base_delay * (2**attempts), self.max_delay)
        self._retry_counts[key] = attempts + 1

        return ErrorDecision(
            action=ErrorAction.RETRY,
            delay_seconds=delay,
            message=f"Timeout, retrying in {delay:.1f}s (attempt {attempts + 1})",
        )

    def _handle_network_error(self, context: ErrorContext) -> ErrorDecision:
        """Handle network errors with retry."""
        key = f"network_{context.iteration}"
        attempts = self._retry_counts.get(key, 0)

        if attempts >= self.max_retries:
            return ErrorDecision(
                action=ErrorAction.ABORT,
                message=f"Network error persisted after {attempts} retries",
            )

        delay = min(self.base_delay * (2**attempts), self.max_delay)
        self._retry_counts[key] = attempts + 1

        return ErrorDecision(
            action=ErrorAction.RETRY,
            delay_seconds=delay,
            message=f"Network error, retrying in {delay:.1f}s (attempt {attempts + 1})",
        )

    def _handle_tool_error(
        self,
        error: Exception,
        context: ErrorContext,
    ) -> ErrorDecision:
        """Handle tool execution errors."""
        error_message = str(error).lower()

        # Some tool errors are recoverable
        if "not found" in error_message or "does not exist" in error_message:
            return ErrorDecision(
                action=ErrorAction.SKIP,
                message=f"Tool {context.tool_name}: resource not found",
                metadata={"error": str(error)},
            )

        if "invalid" in error_message or "invalid argument" in error_message:
            # Bad arguments, don't retry
            return ErrorDecision(
                action=ErrorAction.SKIP,
                message=f"Tool {context.tool_name}: invalid arguments",
                metadata={"error": str(error)},
            )

        # Other tool errors: try once more
        key = f"tool_{context.tool_name}_{context.iteration}"
        attempts = self._retry_counts.get(key, 0)

        if attempts >= 1:
            return ErrorDecision(
                action=ErrorAction.SKIP,
                message=f"Tool {context.tool_name} failed, skipping",
                metadata={"error": str(error)},
            )

        self._retry_counts[key] = attempts + 1
        return ErrorDecision(
            action=ErrorAction.RETRY,
            delay_seconds=self.base_delay,
            message=f"Tool {context.tool_name} failed, retrying",
        )

    def reset(self) -> None:
        """Reset retry counters."""
        self._retry_counts.clear()
